File: Tests/Tests/STPCardTest.m
//
//  STPCardTest.m
//  Stripe
//
//  Created by Saikat Chakrabarti on 11/5/12.
//
//

@import XCTest;

#import "STPCard.h"
#import "STPCard+Private.h"

#import "NSDictionary+Stripe.h"
#import "STPFixtures.h"
#import "STPTestUtils.h"

@interface STPCard ()

@property (nonatomic, assign, readwrite) STPCardBrand brand;

- (void)setLast4:(NSString *)last4;
- (void)setAllResponseFields:(NSDictionary *)allResponseFields;

@end

@interface STPCardTest : XCTestCase

@end

@implementation STPCardTest

#pragma mark - STPCardBrand Tests

// These are only intended to be deprecated publicly.
// When removed from public header, can remove these pragmas
- (void)testBrandFromString {
    XCTAssertEqual([STPCard brandFromString:@"visa"], STPCardBrandVisa);
    XCTAssertEqual([STPCard brandFromString:@"VISA"], STPCardBrandVisa);

    XCTAssertEqual([STPCard brandFromString:@"american express"], STPCardBrandAmex);
    XCTAssertEqual([STPCard brandFromString:@"AMERICAN EXPRESS"], STPCardBrandAmex);

    XCTAssertEqual([STPCard brandFromString:@"mastercard"], STPCardBrandMasterCard);
    XCTAssertEqual([STPCard brandFromString:@"MASTERCARD"], STPCardBrandMasterCard);

    XCTAssertEqual([STPCard brandFromString:@"discover"], STPCardBrandDiscover);
    XCTAssertEqual([STPCard brandFromString:@"DISCOVER"], STPCardBrandDiscover);

    XCTAssertEqual([STPCard brandFromString:@"jcb"], STPCardBrandJCB);
    XCTAssertEqual([STPCard brandFromString:@"JCB"], STPCardBrandJCB);

    XCTAssertEqual([STPCard brandFromString:@"diners club"], STPCardBrandDinersClub);
    XCTAssertEqual([STPCard brandFromString:@"DINERS CLUB"], STPCardBrandDinersClub);

    XCTAssertEqual([STPCard brandFromString:@"unionpay"], STPCardBrandUnionPay);
    XCTAssertEqual([STPCard brandFromString:@"UNIONPAY"], STPCardBrandUnionPay);

    XCTAssertEqual([STPCard brandFromString:@"unknown"], STPCardBrandUnknown);
    XCTAssertEqual([STPCard brandFromString:@"UNKNOWN"], STPCardBrandUnknown);
    
    XCTAssertEqual([STPCard brandFromString:@"garbage"], STPCardBrandUnknown);
    XCTAssertEqual([STPCard brandFromString:@"GARBAGE"], STPCardBrandUnknown);
}

#pragma mark - STPCardFundingType Tests

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated"
// These are only intended to be deprecated publicly.
// When removed from public header, can remove these pragmas

- (void)testFundingFromString {
    XCTAssertEqual([STPCard fundingFromString:@"credit"], STPCardFundingTypeCredit);
    XCTAssertEqual([STPCard fundingFromString:@"CREDIT"], STPCardFundingTypeCredit);

    XCTAssertEqual([STPCard fundingFromString:@"debit"], STPCardFundingTypeDebit);
    XCTAssertEqual([STPCard fundingFromString:@"DEBIT"], STPCardFundingTypeDebit);

    XCTAssertEqual([STPCard fundingFromString:@"prepaid"], STPCardFundingTypePrepaid);
    XCTAssertEqual([STPCard fundingFromString:@"PREPAID"], STPCardFundingTypePrepaid);

    XCTAssertEqual([STPCard fundingFromString:@"other"], STPCardFundingTypeOther);
    XCTAssertEqual([STPCard fundingFromString:@"OTHER"], STPCardFundingTypeOther);

    XCTAssertEqual([STPCard fundingFromString:@"unknown"], STPCardFundingTypeOther);
    XCTAssertEqual([STPCard fundingFromString:@"UNKNOWN"], STPCardFundingTypeOther);

    XCTAssertEqual([STPCard fundingFromString:@"garbage"], STPCardFundingTypeOther);
    XCTAssertEqual([STPCard fundingFromString:@"GARBAGE"], STPCardFundingTypeOther);
}

#pragma clang diagnostic pop

- (void)testStringFromFunding {
    NSArray<NSNumber *> *values = @[
                                    @(STPCardFundingTypeCredit),
                                    @(STPCardFundingTypeDebit),
                                    @(STPCardFundingTypePrepaid),
                                    @(STPCardFundingTypeOther),
                                    ];

    for (NSNumber *fundingNumber in values) {
        STPCardFundingType funding = (STPCardFundingType)[fundingNumber integerValue];
        NSString *string = [STPCard stringFromFunding:funding];

        switch (funding) {
            case STPCardFundingTypeCredit:
                XCTAssertEqualObjects(string, @"credit");
                break;
            case STPCardFundingTypeDebit:
                XCTAssertEqualObjects(string, @"debit");
                break;
            case STPCardFundingTypePrepaid:
                XCTAssertEqualObjects(string, @"prepaid");
                break;
            case STPCardFundingTypeOther:
                XCTAssertNil(string);
                break;
        }
    }
}

#pragma mark -
#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated"
// These tests can ber removed in the future, they should be covered by
// the equivalent response decodeable tests

- (void)testInitWithIDBrandLast4ExpMonthExpYearFunding {
    STPCard *card = [[STPCard alloc] initWithID:@"card_1AVRojEOD54MuFwSxr93QJSx"
                                          brand:STPCardBrandVisa
                                          last4:@"5556"
                                       expMonth:12
                                        expYear:2034
                                        funding:STPCardFundingTypeDebit];
    XCTAssertEqualObjects(card.stripeID, @"card_1AVRojEOD54MuFwSxr93QJSx");
    XCTAssertEqual(card.brand, STPCardBrandVisa);
    XCTAssertEqualObjects(card.last4, @"5556");
    XCTAssertEqual(card.expMonth, (NSUInteger)12);
    XCTAssertEqual(card.expYear, (NSUInteger)2034);
    XCTAssertEqual(card.funding, STPCardFundingTypeDebit);
}

#pragma clang diagnostic pop

- (void)testIsApplePayCard {
    STPCard *card = [STPFixtures card];

    card.allResponseFields = @{};
    XCTAssertFalse(card.isApplePayCard);

    card.allResponseFields = @{@"tokenization_method": @"android_pay"};
    XCTAssertFalse(card.isApplePayCard);

    card.allResponseFields = @{@"tokenization_method": @"apple_pay"};
    XCTAssert(card.isApplePayCard);

    card.allResponseFields = @{@"tokenization_method": @"garbage"};
    XCTAssertFalse(card.isApplePayCard);

    card.allResponseFields = @{@"tokenization_method": @""};
    XCTAssertFalse(card.isApplePayCard);

    // See: https://stripe.com/docs/api#card_object-tokenization_method
}

- (void)testAddressPopulated {
    STPCard *card = [STPFixtures card];
    XCTAssertEqualObjects(card.address.name, @"Jane Austen");
    XCTAssertEqualObjects(card.address.line1, @"123 Fake St");
    XCTAssertEqualObjects(card.address.line2, @"Apt 1");
    XCTAssertEqualObjects(card.address.city, @"Pittsburgh");
    XCTAssertEqualObjects(card.address.state, @"PA");
    XCTAssertEqualObjects(card.address.postalCode, @"19219");
    XCTAssertEqualObjects(card.address.country, @"US");
}

#pragma mark - Equality Tests

- (void)testCardEquals {
    STPCard *card1 = [STPFixtures card];
    STPCard *card2 = [STPFixtures card];

    XCTAssertNotEqual(card1, card2);

    XCTAssertEqualObjects(card1, card1);
    XCTAssertEqualObjects(card1, card2);

    XCTAssertEqual(card1.hash, card1.hash);
    XCTAssertEqual(card1.hash, card2.hash);
}

#pragma mark - Description Tests

- (void)testDescription {
    STPCard *card = [STPFixtures card];
    XCTAssert(card.description);
}

#pragma mark - STPAPIResponseDecodable Tests

- (void)testDecodedObjectFromAPIResponseRequiredFields {
    NSArray<NSString *> *requiredFields = @[
                                            @"id",
                                            @"last4",
                                            @"brand",
                                            @"exp_month",
                                            @"exp_year",
                                            ];

    for (NSString *field in requiredFields) {
        NSMutableDictionary *response = [[STPTestUtils jsonNamed:@"Card"] mutableCopy];
        [response removeObjectForKey:field];

        XCTAssertNil([STPCard decodedObjectFromAPIResponse:response]);
    }

    XCTAssert([STPCard decodedObjectFromAPIResponse:[STPTestUtils jsonNamed:@"Card"]]);
}



- (void)testDecodedObjectFromAPIResponseMapping {
    NSDictionary *response = [STPTestUtils jsonNamed:@"Card"];
    STPCard *card = [STPCard decodedObjectFromAPIResponse:response];

    XCTAssertEqualObjects(card.stripeID, @"card_103kbR2eZvKYlo2CDczLmw4K");

    XCTAssertEqualObjects(card.address.city, @"Pittsburgh");
    XCTAssertEqualObjects(card.address.country, @"US");
    XCTAssertEqualObjects(card.address.line1, @"123 Fake St");
    XCTAssertEqualObjects(card.address.line2, @"Apt 1");
    XCTAssertEqualObjects(card.address.state, @"PA");
    XCTAssertEqualObjects(card.address.postalCode, @"19219");

#pragma clang diagnostic push
#pragma clang diagnostic ignored "-Wdeprecated"

    XCTAssertEqualObjects(card.cardId, @"card_103kbR2eZvKYlo2CDczLmw4K");

    XCTAssertEqualObjects(card.addressCity, @"Pittsburgh");
    XCTAssertEqualObjects(card.addressCountry, @"US");
    XCTAssertEqualObjects(card.addressLine1, @"123 Fake St");
    XCTAssertEqualObjects(card.addressLine2, @"Apt 1");
    XCTAssertEqualObjects(card.addressState, @"PA");
    XCTAssertEqualObjects(card.addressZip, @"19219");

#pragma clang diagnostic pop

    XCTAssertEqual(card.brand, STPCardBrandVisa);
    XCTAssertEqualObjects(card.country, @"US");
    XCTAssertEqualObjects(card.currency, @"usd");
    XCTAssertEqualObjects(card.dynamicLast4, @"5678");
    XCTAssertEqual(card.expMonth, (NSUInteger)5);
    XCTAssertEqual(card.expYear, (NSUInteger)2017);
    XCTAssertEqual(card.funding, STPCardFundingTypeCredit);
    XCTAssertEqualObjects(card.last4, @"4242");
    XCTAssertEqualObjects(card.metadata, @{@"order_id": @"6735"});
    XCTAssertEqualObjects(card.name, @"Jane Austen");

    XCTAssertNotEqual(card.allResponseFields, response);
    XCTAssertEqualObjects(card.allResponseFields, [response stp_dictionaryByRemovingNulls]);
}



#pragma mark - STPSourceProtocol Tests

- (void)testStripeID {
    STPCard *card = [STPFixtures card];
    XCTAssertEqualObjects(card.stripeID, @"card_103kbR2eZvKYlo2CDczLmw4K");
}

#pragma mark - STPPaymentOption Tests

- (void)testImage {
    STPCard *card = [STPFixtures card];
    [self forEachBrand:^(STPCardBrand brand) {
        card.brand = brand;
        XCTAssert([card image]);
    }];
}

- (void)testTemplateImage {
    STPCard *card = [STPFixtures card];
    [self forEachBrand:^(STPCardBrand brand) {
        card.brand = brand;

        XCTAssert([card templateImage]);
    }];
}

- (void)testLabel {
    STPCard *card = [STPCard decodedObjectFromAPIResponse:[STPTestUtils jsonNamed:@"Card"]];
    XCTAssertEqualObjects(card.label, @"Visa 4242");
}

#pragma mark -

- (void)forEachBrand:(void (^)(STPCardBrand brand))block {
    NSArray<NSNumber *> *values = @[
                                    @(STPCardBrandAmex),
                                    @(STPCardBrandDinersClub),
                                    @(STPCardBrandDiscover),
                                    @(STPCardBrandJCB),
                                    @(STPCardBrandMasterCard),
                                    @(STPCardBrandUnionPay),
                                    @(STPCardBrandVisa),
                                    @(STPCardBrandUnknown),
                                    ];

    for (NSNumber *brandNumber in values) {
        block((STPCardBrand)[brandNumber integerValue]);
    }
}

@end

File: Tests/Tests/STPCardTest.swift
//
//  STPCardTest.swift
//  Stripe
//
//  Created by Saikat Chakrabarti on 11/5/12.
//
//

import XCTest

@testable import Stripe

class STPCardTest: XCTestCase {
  // MARK: - STPCardBrand Tests

  // These are only intended to be deprecated publicly.
  // When removed from public header, can remove these pragmas
  func testBrandFromString() {
    XCTAssertEqual(STPCard.brand(from: "visa"), .visa)
    XCTAssertEqual(STPCard.brand(from: "VISA"), .visa)

    XCTAssertEqual(STPCard.brand(from: "american express"), .amex)
    XCTAssertEqual(STPCard.brand(from: "AMERICAN EXPRESS"), .amex)

    XCTAssertEqual(STPCard.brand(from: "mastercard"), .mastercard)
    XCTAssertEqual(STPCard.brand(from: "MASTERCARD"), .mastercard)

    XCTAssertEqual(STPCard.brand(from: "discover"), .discover)
    XCTAssertEqual(STPCard.brand(from: "DISCOVER"), .discover)

    XCTAssertEqual(STPCard.brand(from: "jcb"), .JCB)
    XCTAssertEqual(STPCard.brand(from: "JCB"), .JCB)

    XCTAssertEqual(STPCard.brand(from: "diners club"), .dinersClub)
    XCTAssertEqual(STPCard.brand(from: "DINERS CLUB"), .dinersClub)

    XCTAssertEqual(STPCard.brand(from: "unionpay"), .unionPay)
    XCTAssertEqual(STPCard.brand(from: "UNIONPAY"), .unionPay)

    XCTAssertEqual(STPCard.brand(from: "unknown"), .unknown)
    XCTAssertEqual(STPCard.brand(from: "UNKNOWN"), .unknown)

    XCTAssertEqual(STPCard.brand(from: "garbage"), .unknown)
    XCTAssertEqual(STPCard.brand(from: "GARBAGE"), .unknown)
  }

  // MARK: - STPCardFundingType Tests

  //#pragma clang diagnostic push
  //#pragma clang diagnostic ignored "-Wdeprecated"
  // These are only intended to be deprecated publicly.
  // When removed from public header, can remove these pragmas
  func testFundingFromString() {
    XCTAssertEqual(STPCard.funding(from: "credit"), .credit)
    XCTAssertEqual(STPCard.funding(from: "CREDIT"), .credit)

    XCTAssertEqual(STPCard.funding(from: "debit"), .debit)
    XCTAssertEqual(STPCard.funding(from: "DEBIT"), .debit)

    XCTAssertEqual(STPCard.funding(from: "prepaid"), .prepaid)
    XCTAssertEqual(STPCard.funding(from: "PREPAID"), .prepaid)

    XCTAssertEqual(STPCard.funding(from: "other"), .other)
    XCTAssertEqual(STPCard.funding(from: "OTHER"), .other)

    XCTAssertEqual(STPCard.funding(from: "unknown"), .other)
    XCTAssertEqual(STPCard.funding(from: "UNKNOWN"), .other)

    XCTAssertEqual(STPCard.funding(from: "garbage"), .other)
    XCTAssertEqual(STPCard.funding(from: "GARBAGE"), .other)
  }

  //#pragma clang diagnostic pop
  func testStringFromFunding() {
    let values: [STPCardFundingType] = [
      .credit,
      .debit,
      .prepaid,
      .other,
    ]

    for funding in values {
      let string = STPCard.string(fromFunding: funding)

      switch funding {
      case .credit:
        XCTAssertEqual(string, "credit")
      case .debit:
        XCTAssertEqual(string, "debit")
      case .prepaid:
        XCTAssertEqual(string, "prepaid")
      case .other:
        XCTAssertNil(string)
      default:
        break
      }
    }
  }

  // MARK: -
  //#pragma clang diagnostic push
  //#pragma clang diagnostic ignored "-Wdeprecated"
  // These tests can ber removed in the future, they should be covered by
  // the equivalent response decodeable tests
  func testInitWithIDBrandLast4ExpMonthExpYearFunding() {
    let card = STPCard(
      id: "card_1AVRojEOD54MuFwSxr93QJSx",
      brand: .visa,
      last4: "5556",
      expMonth: 12,
      expYear: 2034,
      funding: .debit)
    XCTAssertEqual(card.stripeID, "card_1AVRojEOD54MuFwSxr93QJSx")
    XCTAssertEqual(card.brand, .visa)
    XCTAssertEqual(card.last4, "5556")
    XCTAssertEqual(card.expMonth, Int(12))
    XCTAssertEqual(card.expYear, Int(2034))
    XCTAssertEqual(card.funding, .debit)
  }

  //#pragma clang diagnostic pop
  func testIsApplePayCard() {
    let card = STPFixtures.card()

    card.allResponseFields = [:]
    XCTAssertFalse(card.isApplePayCard)

    card.allResponseFields = [
      "tokenization_method": "android_pay"
    ]
    XCTAssertFalse(card.isApplePayCard)

    card.allResponseFields = [
      "tokenization_method": "apple_pay"
    ]
    XCTAssertTrue(card.isApplePayCard)

    card.allResponseFields = [
      "tokenization_method": "garbage"
    ]
    XCTAssertFalse(card.isApplePayCard)

    card.allResponseFields = [
      "tokenization_method": ""
    ]
    XCTAssertFalse(card.isApplePayCard)

    // See: https://stripe.com/docs/api#card_object-tokenization_method
  }

  func testAddressPopulated() {
    let card = STPFixtures.card()
    XCTAssertEqual(card.address?.name, "Jane Austen")
    XCTAssertEqual(card.address?.line1, "123 Fake St")
    XCTAssertEqual(card.address?.line2, "Apt 1")
    XCTAssertEqual(card.address?.city, "Pittsburgh")
    XCTAssertEqual(card.address?.state, "PA")
    XCTAssertEqual(card.address?.postalCode, "19219")
    XCTAssertEqual(card.address?.country, "US")
  }

  // MARK: - Equality Tests
  func testCardEquals() {
    let card1 = STPFixtures.card()
    let card2 = STPFixtures.card()

    XCTAssertEqual(card1, card1)
    XCTAssertEqual(card1, card2)

    XCTAssertEqual(card1.hash, card1.hash)
    XCTAssertEqual(card1.hash, card2.hash)
  }

  // MARK: - STPAPIResponseDecodable Tests
  func testDecodedObjectFromAPIResponseRequiredFields() {
    let requiredFields = ["id", "last4", "brand", "exp_month", "exp_year"]

    for field in requiredFields {
      var response = STPTestUtils.jsonNamed("Card")
      response?.removeValue(forKey: field)

      XCTAssertNil(STPCard.decodedObject(fromAPIResponse: response))
    }

    XCTAssert((STPCard.decodedObject(fromAPIResponse: STPTestUtils.jsonNamed("Card")) != nil))
  }

  func testDecodedObjectFromAPIResponseMapping() {
    let response = STPTestUtils.jsonNamed("Card")!
    let card = STPCard.decodedObject(fromAPIResponse: response)!

    XCTAssertEqual(card.stripeID, "card_103kbR2eZvKYlo2CDczLmw4K")

    XCTAssertEqual(card.address?.city, "Pittsburgh")
    XCTAssertEqual(card.address?.country, "US")
    XCTAssertEqual(card.address?.line1, "123 Fake St")
    XCTAssertEqual(card.address?.line2, "Apt 1")
    XCTAssertEqual(card.address?.state, "PA")
    XCTAssertEqual(card.address?.postalCode, "19219")

    //#pragma clang diagnostic push
    //#pragma clang diagnostic ignored "-Wdeprecated"

    XCTAssertEqual(card.cardId, "card_103kbR2eZvKYlo2CDczLmw4K")

    XCTAssertEqual(card.addressCity, "Pittsburgh")
    XCTAssertEqual(card.addressCountry, "US")
    XCTAssertEqual(card.addressLine1, "123 Fake St")
    XCTAssertEqual(card.addressLine2, "Apt 1")
    XCTAssertEqual(card.addressState, "PA")
    XCTAssertEqual(card.addressZip, "19219")
    XCTAssertNil(card.metadata)

    //#pragma clang diagnostic pop

    XCTAssertEqual(card.brand, .visa)
    XCTAssertEqual(card.country, "US")
    XCTAssertEqual(card.currency, "usd")
    XCTAssertEqual(card.dynamicLast4, "5678")
    XCTAssertEqual(card.expMonth, Int(5))
    XCTAssertEqual(card.expYear, Int(2017))
    XCTAssertEqual(card.funding, .credit)
    XCTAssertEqual(card.last4, "4242")
    XCTAssertEqual(card.name, "Jane Austen")

    XCTAssertEqual(card.allResponseFields as NSDictionary, response as NSDictionary)
  }

  // MARK: - STPSourceProtocol Tests
  func testStripeID() {
    let card = STPFixtures.card()
    XCTAssertEqual(card.stripeID, "card_103kbR2eZvKYlo2CDczLmw4K")
  }

  // MARK: - STPPaymentOption Tests
  func testLabel() {
    let card = STPCard.decodedObject(fromAPIResponse: STPTestUtils.jsonNamed("Card"))!
    XCTAssertEqual(card.label, "Visa 4242")
  }

  // MARK: -
  func forEachBrand(_ block: @escaping (_ brand: STPCardBrand) -> Void) {
    let values: [STPCardBrand] = [
      .amex,
      .dinersClub,
      .discover,
      .JCB,
      .mastercard,
      .unionPay,
      .visa,
      .unknown,
    ]

    for brand in values {
      block(brand)
    }
  }
}
